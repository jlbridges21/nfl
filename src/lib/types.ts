// Type definitions for the NFL Game Predictor

export interface Team {
  code: string;
  name: string;
  city: string;
  color: string;
}

export interface GamePrediction {
  awayTeam: Team;
  homeTeam: Team;
  awayScore: number;
  homeScore: number;
  spread: number;
  total: number;
  awayWinProbability: number;
  homeWinProbability: number;
  confidence: number;
}

export interface PredictionFactor {
  id: string;
  label: string;
  icon: string;
  contribution: number;
  description: string;
  color: 'positive' | 'negative' | 'neutral';
}

export interface MatchupSelection {
  awayTeam: string | null;
  homeTeam: string | null;
}

// Loading states
export type LoadingState = 'idle' | 'loading' | 'success' | 'error';

// Theme types
export type Theme = 'light' | 'dark' | 'system';
