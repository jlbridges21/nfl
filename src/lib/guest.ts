/**
 * Gets or creates a persistent guest device ID
 * Priority: httpOnly cookie → localStorage → generate new
 * Stores in both localStorage and httpOnly cookie for persistence across auth events
 */
export async function getOrCreateDeviceId(): Promise<string> {
  const STORAGE_KEY = 'guestDeviceId';

  try {
    // First, try to sync with server cookie via our device API
    const syncResponse = await fetch('/api/guest/device', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ action: 'get' }),
    });

    if (syncResponse.ok) {
      const { deviceId: cookieId } = await syncResponse.json();
      if (cookieId && cookieId.length > 0) {
        // Update localStorage to match cookie
        try {
          if (typeof window !== 'undefined' && window.localStorage) {
            localStorage.setItem(STORAGE_KEY, cookieId);
          }
        } catch (error) {
          console.warn('Failed to update localStorage with cookie value:', error);
        }
        return cookieId;
      }
    }
  } catch (error) {
    console.warn('Failed to sync with server cookie:', error);
  }

  try {
    // Fallback to localStorage
    if (typeof window !== 'undefined' && window.localStorage) {
      const existingId = localStorage.getItem(STORAGE_KEY);
      if (existingId && existingId.length > 0) {
        // Set the cookie to match localStorage
        try {
          await fetch('/api/guest/device', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ action: 'set', deviceId: existingId }),
          });
        } catch (error) {
          console.warn('Failed to set cookie from localStorage:', error);
        }
        return existingId;
      }
    }
  } catch (error) {
    console.warn('localStorage not available for guest device ID:', error);
  }

  // Generate new UUID v4 using crypto API
  const newId = crypto.randomUUID();

  try {
    // Store in localStorage
    if (typeof window !== 'undefined' && window.localStorage) {
      localStorage.setItem(STORAGE_KEY, newId);
    }
  } catch (error) {
    console.warn('Failed to store guest device ID in localStorage:', error);
  }

  try {
    // Set httpOnly cookie via server route
    await fetch('/api/guest/device', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ action: 'set', deviceId: newId }),
    });
  } catch (error) {
    console.warn('Failed to set guest device ID cookie:', error);
  }

  return newId;
}

/**
 * Synchronous version for backwards compatibility
 * Uses localStorage only - async version should be preferred
 */
export function getGuestDeviceId(): string {
  const STORAGE_KEY = 'guestDeviceId';

  try {
    // Try to get existing ID from localStorage
    if (typeof window !== 'undefined' && window.localStorage) {
      const existingId = localStorage.getItem(STORAGE_KEY);
      if (existingId && existingId.length > 0) {
        return existingId;
      }
    }
  } catch (error) {
    console.warn('localStorage not available for guest device ID:', error);
  }

  // Generate new UUID v4 using crypto API
  const newId = crypto.randomUUID();

  try {
    // Try to store in localStorage
    if (typeof window !== 'undefined' && window.localStorage) {
      localStorage.setItem(STORAGE_KEY, newId);
    }
  } catch (error) {
    console.warn('Failed to store guest device ID in localStorage:', error);
  }

  return newId;
}
