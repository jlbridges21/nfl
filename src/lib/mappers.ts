import type { SeasonStatsRow } from "../types/db";
import type { TeamStats } from "../types/model";

/**
 * Helper function to safely convert null/undefined to zero
 * Used internally by the mapper functions
 */
export function nullToZero(n: number | null | undefined): number {
  return n ?? 0;
}

/**
 * Maps a raw SeasonStatsRow from the database to a compact TeamStats object
 * for use by the scoring engine. All numeric values are coerced to numbers
 * with null/undefined values defaulting to 0.
 */
export function toTeamStats(row: SeasonStatsRow): TeamStats {
  return {
    // Core identifier
    year: nullToZero(typeof row.year === "string" ? parseFloat(row.year) : row.year),

    // Offense - Yards per Game
    yardsPerGameSeason: nullToZero(typeof row.yardsPerGame === "string" ? parseFloat(row.yardsPerGame) : row.yardsPerGame),
    yardsPerGameLast3: nullToZero(typeof row.yardsPerGameLast3 === "string" ? parseFloat(row.yardsPerGameLast3) : row.yardsPerGameLast3),
    yardsPerGameLast1: nullToZero(typeof row.yardsPerGameLast1 === "string" ? parseFloat(row.yardsPerGameLast1) : row.yardsPerGameLast1),
    yardsPerGameHome: nullToZero(typeof row.yardsPerGameHome === "string" ? parseFloat(row.yardsPerGameHome) : row.yardsPerGameHome),
    yardsPerGameAway: nullToZero(typeof row.yardsPerGameAway === "string" ? parseFloat(row.yardsPerGameAway) : row.yardsPerGameAway),

    // Offense - Points per Game
    pointsPerGameSeason: nullToZero(typeof row.pointsPerGame === "string" ? parseFloat(row.pointsPerGame) : row.pointsPerGame),
    pointsPerGameLast3: nullToZero(typeof row.pointsPerGameLast3 === "string" ? parseFloat(row.pointsPerGameLast3) : row.pointsPerGameLast3),
    pointsPerGameLast1: nullToZero(typeof row.pointsPerGameLast1 === "string" ? parseFloat(row.pointsPerGameLast1) : row.pointsPerGameLast1),
    pointsPerGameHome: nullToZero(typeof row.pointsPerGameHome === "string" ? parseFloat(row.pointsPerGameHome) : row.pointsPerGameHome),
    pointsPerGameAway: nullToZero(typeof row.pointsPerGameAway === "string" ? parseFloat(row.pointsPerGameAway) : row.pointsPerGameAway),

    // Offense - Touchdowns per Game (season only for now)
    touchdownsPerGameSeason: nullToZero(typeof row.touchdownsPerGame === "string" ? parseFloat(row.touchdownsPerGame) : row.touchdownsPerGame),

    // Defense - Allowed per Game
    defensiveYardsAllowedSeason: nullToZero(typeof row.opponentYardsPerGame === "string" ? parseFloat(row.opponentYardsPerGame) : row.opponentYardsPerGame),
    defensivePointsAllowedSeason: nullToZero(typeof row.opponentPointsPerGame === "string" ? parseFloat(row.opponentPointsPerGame) : row.opponentPointsPerGame),

    // Efficiency - Yards per Point
    yardsPerPointSeason: nullToZero(typeof row.yardsPerPoint === "string" ? parseFloat(row.yardsPerPoint) : row.yardsPerPoint),
    yardsPerPointLast3: nullToZero(typeof row.yardsPerPointLast3 === "string" ? parseFloat(row.yardsPerPointLast3) : row.yardsPerPointLast3),
    yardsPerPointLast1: nullToZero(typeof row.yardsPerPointLast1 === "string" ? parseFloat(row.yardsPerPointLast1) : row.yardsPerPointLast1),
    yardsPerPointHome: nullToZero(typeof row.yardsPerPointHome === "string" ? parseFloat(row.yardsPerPointHome) : row.yardsPerPointHome),
    yardsPerPointAway: nullToZero(typeof row.yardsPerPointAway === "string" ? parseFloat(row.yardsPerPointAway) : row.yardsPerPointAway),

    // Power Ratings (optional, default to 0)
    fpiOverall: nullToZero(typeof row.fpi === "string" ? parseFloat(row.fpi) : row.fpi),
    fpiOffense: nullToZero(typeof row.fpiOffense === "string" ? parseFloat(row.fpiOffense) : row.fpiOffense),
    fpiDefense: nullToZero(typeof row.fpiDefense === "string" ? parseFloat(row.fpiDefense) : row.fpiDefense),
  };
}
