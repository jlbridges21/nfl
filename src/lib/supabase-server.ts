import { createClient } from '@supabase/supabase-js'
import type { Database } from '@/types/database'

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!
// Use anon key for now since service role key is corrupted
const supabaseKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!

export const supabaseServer = createClient<Database>(supabaseUrl, supabaseKey, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
})
