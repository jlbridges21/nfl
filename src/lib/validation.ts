import { z } from "zod";
import type { TeamsRow, SeasonStatsRow } from "../types/db";

/**
 * Helper function to coerce various input types to numbers
 * Handles null, undefined, empty strings, and string numbers
 */
const coerceToNumber = z
  .union([z.string(), z.number(), z.null(), z.undefined()])
  .transform((val) => {
    if (val === null || val === undefined || val === "") return 0;
    const num = typeof val === "string" ? parseFloat(val) : val;
    return isNaN(num) ? 0 : num;
  });

/**
 * Zod schema for Teams table validation
 */
export const ZTeamsRow = z.object({
  id: z.string(),
  name: z.string(),
  abbreviation: z.string(),
  conference: z.enum(["AFC", "NFC"]),
  division: z.enum(["East", "North", "South", "West"]),
  logoUrl: z.string().nullable(),
  primaryColor: z.string().nullable(),
  secondaryColor: z.string().nullable(),
}) satisfies z.ZodType<TeamsRow>;

/**
 * Zod schema for SeasonStats table validation
 * All numeric fields use coerceToNumber to handle null/string inputs
 */
export const ZSeasonStatsRow = z.object({
  // Core identifiers
  id: z.string(),
  teamId: z.string(),
  year: coerceToNumber,

  // Offense - Yards per Game
  yardsPerGame: coerceToNumber,
  yardsPerGameLast3: coerceToNumber,
  yardsPerGameLast1: coerceToNumber,
  yardsPerGameHome: coerceToNumber,
  yardsPerGameAway: coerceToNumber,

  // Offense - Points per Game
  pointsPerGame: coerceToNumber,
  pointsPerGameLast3: coerceToNumber,
  pointsPerGameLast1: coerceToNumber,
  pointsPerGameHome: coerceToNumber,
  pointsPerGameAway: coerceToNumber,

  // Offense - Touchdowns per Game
  touchdownsPerGame: coerceToNumber,
  touchdownsPerGameLast3: coerceToNumber,
  touchdownsPerGameLast1: coerceToNumber,
  touchdownsPerGameHome: coerceToNumber,
  touchdownsPerGameAway: coerceToNumber,

  // Offense - Passing Yards per Game
  passingYardsPerGame: coerceToNumber,
  passingYardsPerGameLast3: coerceToNumber,
  passingYardsPerGameLast1: coerceToNumber,
  passingYardsPerGameHome: coerceToNumber,
  passingYardsPerGameAway: coerceToNumber,

  // Offense - Rushing Yards per Game
  rushingYardsPerGame: coerceToNumber,
  rushingYardsPerGameLast3: coerceToNumber,
  rushingYardsPerGameLast1: coerceToNumber,
  rushingYardsPerGameHome: coerceToNumber,
  rushingYardsPerGameAway: coerceToNumber,

  // Offense - Red Zone Scores per Game (TD only)
  redZoneScoresPerGameTdOnly: coerceToNumber,
  redZoneScoresPerGameTdOnlyLast3: coerceToNumber,
  redZoneScoresPerGameTdOnlyLast1: coerceToNumber,
  redZoneScoresPerGameTdOnlyHome: coerceToNumber,
  redZoneScoresPerGameTdOnlyAway: coerceToNumber,

  // Defense - Opponent Yards per Game
  opponentYardsPerGame: coerceToNumber,
  opponentYardsPerGameLast3: coerceToNumber,
  opponentYardsPerGameLast1: coerceToNumber,
  opponentYardsPerGameHome: coerceToNumber,
  opponentYardsPerGameAway: coerceToNumber,

  // Defense - Opponent Points per Game
  opponentPointsPerGame: coerceToNumber,
  opponentPointsPerGameLast3: coerceToNumber,
  opponentPointsPerGameLast1: coerceToNumber,
  opponentPointsPerGameHome: coerceToNumber,
  opponentPointsPerGameAway: coerceToNumber,

  // Defense - Opponent Passing Yards per Game
  opponentPassingYardsPerGame: coerceToNumber,
  opponentPassingYardsPerGameLast3: coerceToNumber,
  opponentPassingYardsPerGameLast1: coerceToNumber,
  opponentPassingYardsPerGameHome: coerceToNumber,
  opponentPassingYardsPerGameAway: coerceToNumber,

  // Defense - Opponent Rushing Yards per Game
  opponentRushingYardsPerGame: coerceToNumber,
  opponentRushingYardsPerGameLast3: coerceToNumber,
  opponentRushingYardsPerGameLast1: coerceToNumber,
  opponentRushingYardsPerGameHome: coerceToNumber,
  opponentRushingYardsPerGameAway: coerceToNumber,

  // Defense - Opponent Red Zone Scores per Game (TDs only)
  opponentRedZoneScoresPerGameTdsOnly: coerceToNumber,
  opponentRedZoneScoresPerGameTdsOnlyLast3: coerceToNumber,
  opponentRedZoneScoresPerGameTdsOnlyLast1: coerceToNumber,
  opponentRedZoneScoresPerGameTdsOnlyHome: coerceToNumber,
  opponentRedZoneScoresPerGameTdsOnlyAway: coerceToNumber,

  // Efficiency - Average Scoring Margin
  averageScoringMargin: coerceToNumber,
  averageScoringMarginLast3: coerceToNumber,
  averageScoringMarginLast1: coerceToNumber,
  averageScoringMarginHome: coerceToNumber,
  averageScoringMarginAway: coerceToNumber,

  // Efficiency - Yards per Point
  yardsPerPoint: coerceToNumber,
  yardsPerPointLast3: coerceToNumber,
  yardsPerPointLast1: coerceToNumber,
  yardsPerPointHome: coerceToNumber,
  yardsPerPointAway: coerceToNumber,

  // Power Ratings (optional)
  fpi: coerceToNumber,
  fpiOffense: coerceToNumber,
  fpiDefense: coerceToNumber,
}) satisfies z.ZodType<SeasonStatsRow>;

/**
 * Helper function to validate and parse Teams row data
 */
export function asTeamsRow(input: unknown): TeamsRow {
  return ZTeamsRow.parse(input);
}

/**
 * Helper function to validate and parse SeasonStats row data
 */
export function asSeasonStatsRow(input: unknown): SeasonStatsRow {
  return ZSeasonStatsRow.parse(input);
}
