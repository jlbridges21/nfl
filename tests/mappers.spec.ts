import { describe, it, expect, beforeEach } from "vitest";
import { toTeamStats, nullToZero } from "../src/lib/mappers";
import { asSeasonStatsRow } from "../src/lib/validation";
import type { TeamStats } from "../src/types/model";
import fixtureData from "../src/fixtures/seasonStats.example.json";

describe("nullToZero helper", () => {
  it("should convert null to 0", () => {
    expect(nullToZero(null)).toBe(0);
  });

  it("should convert undefined to 0", () => {
    expect(nullToZero(undefined)).toBe(0);
  });

  it("should preserve valid numbers", () => {
    expect(nullToZero(42)).toBe(42);
    expect(nullToZero(0)).toBe(0);
    expect(nullToZero(-5.5)).toBe(-5.5);
  });
});

describe("Zod validation with fixture", () => {
  it("should validate and coerce the fixture data correctly", () => {
    const validatedRow = asSeasonStatsRow(fixtureData);
    
    // Test that string numbers are coerced to numbers
    expect(typeof validatedRow.year).toBe("number");
    expect(validatedRow.year).toBe(2024);
    
    expect(typeof validatedRow.yardsPerGame).toBe("number");
    expect(validatedRow.yardsPerGame).toBe(385.2);
    
    // Test that nulls are coerced to 0
    expect(validatedRow.pointsPerGameLast1).toBe(0);
    expect(validatedRow.touchdownsPerGameHome).toBe(0);
    expect(validatedRow.passingYardsPerGameAway).toBe(0);
    
    // Test that existing numbers are preserved
    expect(validatedRow.pointsPerGame).toBe(28.4);
    expect(validatedRow.touchdownsPerGame).toBe(3.2);
  });
});

describe("toTeamStats mapper", () => {
  let validatedRow: ReturnType<typeof asSeasonStatsRow>;
  let teamStats: TeamStats;

  beforeEach(() => {
    validatedRow = asSeasonStatsRow(fixtureData);
    teamStats = toTeamStats(validatedRow);
  });

  it("should map core identifier correctly", () => {
    expect(teamStats.year).toBe(2024);
  });

  it("should map offensive yards per game correctly", () => {
    expect(teamStats.yardsPerGameSeason).toBe(385.2);
    expect(teamStats.yardsPerGameLast3).toBe(412.7);
    expect(teamStats.yardsPerGameLast1).toBe(445.0);
    expect(teamStats.yardsPerGameHome).toBe(398.5);
    expect(teamStats.yardsPerGameAway).toBe(371.9);
  });

  it("should map offensive points per game correctly", () => {
    expect(teamStats.pointsPerGameSeason).toBe(28.4);
    expect(teamStats.pointsPerGameLast3).toBe(31.7);
    expect(teamStats.pointsPerGameLast1).toBe(0); // was null in fixture
    expect(teamStats.pointsPerGameHome).toBe(30.2);
    expect(teamStats.pointsPerGameAway).toBe(26.6);
  });

  it("should map touchdowns per game correctly", () => {
    expect(teamStats.touchdownsPerGameSeason).toBe(3.2);
  });

  it("should map defensive stats correctly", () => {
    expect(teamStats.defensiveYardsAllowedSeason).toBe(342.7);
    expect(teamStats.defensivePointsAllowedSeason).toBe(22.1);
  });

  it("should map efficiency stats correctly", () => {
    expect(teamStats.yardsPerPointSeason).toBe(13.6);
    expect(teamStats.yardsPerPointLast3).toBe(13.0);
    expect(teamStats.yardsPerPointLast1).toBe(12.8);
    expect(teamStats.yardsPerPointHome).toBe(13.2);
    expect(teamStats.yardsPerPointAway).toBe(0); // was null in fixture
  });

  it("should map power ratings correctly", () => {
    expect(teamStats.fpiOverall).toBe(8.2);
    expect(teamStats.fpiOffense).toBe(12.1);
    expect(teamStats.fpiDefense).toBe(4.3);
  });

  it("should handle null values by converting to 0", () => {
    // Test specific fields that were null in the fixture
    expect(teamStats.pointsPerGameLast1).toBe(0);
    expect(teamStats.yardsPerPointAway).toBe(0);
  });

  it("should handle string numbers correctly", () => {
    // Test specific fields that were strings in the fixture
    expect(teamStats.yardsPerGameSeason).toBe(385.2); // was "385.2"
    expect(teamStats.pointsPerGameLast3).toBe(31.7); // was "31.7"
    expect(teamStats.fpiOverall).toBe(8.2); // was "8.2"
  });

  it("should produce a complete TeamStats object with all required fields", () => {
    const expectedFields = [
      "year",
      "yardsPerGameSeason", "yardsPerGameLast3", "yardsPerGameLast1", "yardsPerGameHome", "yardsPerGameAway",
      "pointsPerGameSeason", "pointsPerGameLast3", "pointsPerGameLast1", "pointsPerGameHome", "pointsPerGameAway",
      "touchdownsPerGameSeason",
      "defensiveYardsAllowedSeason", "defensivePointsAllowedSeason",
      "yardsPerPointSeason", "yardsPerPointLast3", "yardsPerPointLast1", "yardsPerPointHome", "yardsPerPointAway",
      "fpiOverall", "fpiOffense", "fpiDefense"
    ];

    expectedFields.forEach(field => {
      expect(teamStats).toHaveProperty(field);
      expect(typeof teamStats[field as keyof TeamStats]).toBe("number");
    });
  });

  it("should match expected snapshot", () => {
    expect(teamStats).toMatchSnapshot();
  });
});

describe("Field mapping verification", () => {
  it("should correctly map specific database columns to TeamStats fields", () => {
    const validatedRow = asSeasonStatsRow(fixtureData);
    const teamStats = toTeamStats(validatedRow);

    // Verify explicit field mappings as specified in requirements
    expect(teamStats.yardsPerGameSeason).toBe(validatedRow.yardsPerGame);
    expect(teamStats.yardsPerGameLast3).toBe(validatedRow.yardsPerGameLast3);
    expect(teamStats.yardsPerGameLast1).toBe(validatedRow.yardsPerGameLast1);
    expect(teamStats.yardsPerGameHome).toBe(validatedRow.yardsPerGameHome);
    expect(teamStats.yardsPerGameAway).toBe(validatedRow.yardsPerGameAway);

    expect(teamStats.pointsPerGameSeason).toBe(validatedRow.pointsPerGame);
    expect(teamStats.pointsPerGameLast3).toBe(validatedRow.pointsPerGameLast3);
    expect(teamStats.pointsPerGameLast1).toBe(validatedRow.pointsPerGameLast1);
    expect(teamStats.pointsPerGameHome).toBe(validatedRow.pointsPerGameHome);
    expect(teamStats.pointsPerGameAway).toBe(validatedRow.pointsPerGameAway);

    expect(teamStats.touchdownsPerGameSeason).toBe(validatedRow.touchdownsPerGame);
    expect(teamStats.defensiveYardsAllowedSeason).toBe(validatedRow.opponentYardsPerGame);
    expect(teamStats.defensivePointsAllowedSeason).toBe(validatedRow.opponentPointsPerGame);

    expect(teamStats.yardsPerPointSeason).toBe(validatedRow.yardsPerPoint);
    expect(teamStats.yardsPerPointLast3).toBe(validatedRow.yardsPerPointLast3);
    expect(teamStats.yardsPerPointLast1).toBe(validatedRow.yardsPerPointLast1);
    expect(teamStats.yardsPerPointHome).toBe(validatedRow.yardsPerPointHome);
    expect(teamStats.yardsPerPointAway).toBe(validatedRow.yardsPerPointAway);

    expect(teamStats.fpiOverall).toBe(validatedRow.fpi);
    expect(teamStats.fpiOffense).toBe(validatedRow.fpiOffense);
    expect(teamStats.fpiDefense).toBe(validatedRow.fpiDefense);
  });
});
