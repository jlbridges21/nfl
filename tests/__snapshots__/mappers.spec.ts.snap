// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toTeamStats mapper > should match expected snapshot 1`] = `
{
  "defensivePointsAllowedSeason": 22.1,
  "defensiveYardsAllowedSeason": 342.7,
  "fpiDefense": 4.3,
  "fpiOffense": 12.1,
  "fpiOverall": 8.2,
  "pointsPerGameAway": 26.6,
  "pointsPerGameHome": 30.2,
  "pointsPerGameLast1": 0,
  "pointsPerGameLast3": 31.7,
  "pointsPerGameSeason": 28.4,
  "touchdownsPerGameSeason": 3.2,
  "yardsPerGameAway": 371.9,
  "yardsPerGameHome": 398.5,
  "yardsPerGameLast1": 445,
  "yardsPerGameLast3": 412.7,
  "yardsPerGameSeason": 385.2,
  "yardsPerPointAway": 0,
  "yardsPerPointHome": 13.2,
  "yardsPerPointLast1": 12.8,
  "yardsPerPointLast3": 13,
  "yardsPerPointSeason": 13.6,
  "year": 2024,
}
`;
